README
Any::Template v1.013

(c) BBC 2004. This program is free software; you can redistribute it and/or modify it under the GNU GPL.

See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt 

QUICK START:

Install Any::Template by unpacking the tarball and running the following
commands in the source directory:

    perl Makefile.PL
    make
    make test
    make install
    Then delete the source directory tree since it's no longer needed.

    run 'perldoc Any::Template' to read the full documentation.

BACKENDS INCLUDED WITH THIS DISTRIBUTION:

    TemplateToolkit
    HTML::Template
    Text::Template
    Text::MicroMason

the Template languages currently supported are:

    Template (Template Toolkit)
    HTML::Template (directly and also the emulation via Text::MicroMason)
    Text::Template (directly and also the emulation via Text::MicroMason)
    HTML::Mason (emulated via Text::MicroMason)
    Embperl (emulated via Text::MicroMason)
    Apache::ASP (emulated via Text::MicroMason)


Changes
Mon Aug  1 11:42:57 2005 - 1.013
	- Remove orphaned IO::Scalar dependencies which weren't listed as prereqs
	Added some other non-core (as of 5.004) prereqs

Sun May  8 16:12:47 2005 - 1.012
	- Added missing prereqs for unit tests



lib/Any/Template/Backend.pm
package Any::Template::Backend;

use vars qw($VERSION);
$VERSION = sprintf"%d.%03d", q$Revision: 1.7 $ =~ /: (\d+)\.(\d+)/;

sub process_to_file {
	my ($self, $data, $filepath) = @_;
	local *FH;
	open (FH, ">$filepath") or die("Unable to open $filepath - $!");
	$self->process_to_filehandle($data, \*FH);
	close FH;
}

sub process_to_filehandle {
	my ($self, $data, $fh) = @_;
	my $buffer;
	$self->process_to_string($data, \$buffer);
	print $fh $buffer;	
}

sub process_to_sub {
	my ($self, $data, $coderef) = @_;
	my $buffer;
	$self->process_to_string($data, \$buffer);
	return $coderef->($buffer);	
}

sub preprocess {
	return $_[1]; #no-op by default	
}

1;

=head1 NAME

Any::Template::Backend - base class for implementing backends for Any::Template

=head1 SYNOPSIS

	package Any::Template::Backend::MyTemplate;
	
	use MyTemplate;
	use Any::Template::Backend;
	use vars qw(@ISA);
	@ISA = qw(Any::Template::Backend);
	
	sub new {
		...		
	}

=head1 DESCRIPTION

This exists purely to be inherited from.  It provides some default implementations of sending output to
different sinks based on the lowest common denominator of returning data to a string or filehandle.
You can override these implementations with more efficient ones where the templating engine provides them
natively.

=head1 API FOR BACKEND MODULES

=over 4

=item $o = new Any::Template::Backend::MyBackend(\%options);

You MUST supply a constructor.  There is no default contructor to inherit.
The constructor will need to marshal the options and create the backend template object from them.
Example implementations can be found in backend classes included with the distribution.

=item $templating_engine = $o->native_object()

Returns the underlying native template object.  You SHOULD supply this method.
Although accessing the underlying object defeats the point of Any::Template,
a valid use is in refactoring code, where dependencies on a particular engine's API
can be eradicated in iterations.

=item $data = $o->preprocess($data)

You CAN supply a method to preprocess the data structure before it's handed off to one of the process methods listed below.
Typically you might use this to remove some values from the data structure (e.g. globrefs) that a template backend
might not be able to handle.
The default implementation returns $data unmodified.

=item $o->process_to_string($data, $scalar_ref)

You MUST supply this method.
Example implementations can be found in backend classes included with the distribution.

=item $o->process_to_filehandle($data, $fh)

You CAN supply this method.  If you don't, output will be collected in a string and sent to the filehandle.

=item $o->process_to_sub($data, \&code)

You CAN supply this method.  If you don't, output will be collected in a string and sent to the sub.

=item $o->process_to_file($data, $filename)

You CAN supply this method.  If you don't, a filehandle will be opened against this file,
and process_to_filehandle will be used to populate it.

=back

=head1 VERSION

$Revision: 1.7 $ on $Date: 2005/03/23 12:13:07 $ by $Author: tonyh $

=head1 AUTHOR

John Alden <cpan _at_ bbc _dot_ co _dot_ uk>

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.

See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt 

=cut


lib/Any/Template/Backend/TemplateToolkit.pm
package Any::Template::Backend::TemplateToolkit;

use strict;
use Template;
use Any::Template::Backend;
use vars qw(@ISA);
@ISA = qw(Any::Template::Backend);

use vars qw($VERSION);
$VERSION = sprintf"%d.%03d", q$Revision: 1.6 $ =~ /: (\d+)\.(\d+)/;

sub new {
	my ($class, $options) = @_;
	my $self = bless {}, $class;
	my $marshalled = $self->_marshall_options($options);
	$self->{engine} = new Template(%$marshalled);
	return $self;		
}

sub native_object {
	my $self = shift;
	return $self->{engine};	
}

sub process_to_string {
	my ($self, $data, $ref_buffer) = @_;
	$$ref_buffer = '';
	TRACE("Input", $self->{input});
	$self->{engine}->process($self->{input}, $data, $ref_buffer) or die();
}

sub process_to_filehandle {
	my ($self, $data, $fh) = @_;
	TRACE("Input", $self->{input});
	$self->{engine}->process($self->{input}, $data, $fh) or die();
}

sub process_to_file {
	my ($self, $data, $filepath) = @_;
	TRACE("Input", $self->{input});
	$self->{engine}->process($self->{input}, $data, $filepath) or die();
}

sub process_to_sub {
	my ($self, $data, $coderef) = @_;
	return $self->{engine}->process($self->{input}, $data, $coderef);
}

#
# This marshalls the Any::Template ctor options into the form required for Template-Toolkit
#
sub _marshall_options {
	my $self = shift;
	my $at_options = shift;
	my %tt_options = %{$at_options->{Options}};
	if(exists $at_options->{String}) {
		$self->{input} = \$at_options->{String};
	}
	elsif(exists $at_options->{Filehandle}) {
		$self->{input} = $at_options->{Filehandle}; 		
	}
	elsif(exists $at_options->{Filename}) {
		$self->{input} = $at_options->{Filename};
	}
	else {
		die("No Filename, Filehandle or String");	
	}
	return \%tt_options;
}

#Log::Trace stubs
sub TRACE{}
sub DUMP{}

1;


=head1 NAME

Any::Template::Backend::TemplateToolkit - Any::Template backend for Template Toolkit

=head1 SYNOPSIS

	use Any::Template;
	my $template = new Any::Template(
		Backend => 'TemplateToolkit',
		Options => {
			'POST_CHOMP' => 1, #Template ctor options
		},
		File => 'page.tmpl'
	);	
	my $output = $template->process($data);

=head1 DESCRIPTION

All input and output methods are implemented using Template Toolkit's native features so they should
all be pretty efficient.

=head1 SEE ALSO

L<Any::Template>, L<Any::Template::Backend>, L<Template>

=head1 VERSION

$Revision: 1.6 $ on $Date: 2005/05/08 18:25:16 $ by $Author: johna $

=head1 AUTHOR

John Alden <cpan _at_ bbc _dot_ co _dot_ uk>

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.

See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt 

=cut

lib/Any/Template/Backend/Text/MicroMason.pm
package Any::Template::Backend::Text::MicroMason;

use strict;
use Text::MicroMason;
use Any::Template::Backend;
use vars qw(@ISA);
@ISA = qw(Any::Template::Backend);

use vars qw($VERSION);
$VERSION = sprintf"%d.%03d", q$Revision: 1.6 $ =~ /: (\d+)\.(\d+)/;

sub new {
	my ($class, $options) = @_;
	my $attributes = $options->{Options}{Attributes} || {};
	my $mixins = $options->{Options}{Mixins} || [];
	my $compile_options = _compile_options($options);
	my $factory = new Text::MicroMason(@$mixins, %$attributes);
	my $self = {
		engine => $factory->compile(%$compile_options)
	};
	return bless($self, $class);
}

sub native_object {
	my $self = shift;
	return $self->{engine};	
}

sub process_to_string {
	my ($self, $data, $ref_buffer) = @_;
	die("data must be a hashref") unless(ref $data eq "HASH");
	$$ref_buffer = $self->{engine}->(%$data);
	return 1;
}

#
# This marshalls the Any::Template ctor options into the form required for Text::MicroMason
#
sub _compile_options {
	my $at_options = shift;
	my %options;
	if(exists $at_options->{String}) {
		$options{text} = $at_options->{String};
	}
	elsif(exists $at_options->{Filename}) {
		$options{file} = $at_options->{Filename};
	}
	elsif(exists $at_options->{Filehandle}) {
		$options{text} = _fh_to_string($at_options->{Filehandle});
	}
	return \%options;
}

sub _fh_to_string {
	my $fh = shift;
	local $/ = undef;
	$fh = $$fh if(ref $fh eq 'GLOB');
	my $string = <$fh>;
	return $string;
}

sub DUMP {}
sub TRACE {}

1;

=head1 NAME

Any::Template::Backend::Text::MicroMason - Any::Template backend for Text::MicroMason

=head1 SYNOPSIS

	use Any::Template;
	my $template = new Any::Template(
		Backend => 'Text::MicroMason',
		Options => {
			Attributes => {global_vars => 1},      #MicroMason %attribs
			Mixins => [qw(-HTMLTemplate -Filters)] #Specify mixins
		},
		File => 'page.tmpl'
	);	
	my $output = $template->process($data);

=head1 DESCRIPTION

Attributes may be passed to Text::MicroMason in the {Options}{Attributes} key.  
The {Options}{Mixins} key is used to pass mixins to Text::MicroMason.  

Inputs from a file and string are provided natively by Text::MicroMason.
Input from a filehandle uses the default implementation (which is to read into a string).

Output to filehandle is based on the default implementation of capturing output in a string and writing
this to a filehandle, so watch out for this if your output is very large.  Output to a file uses the native
Text::MicroMason implementation.

Output to a coderef uses the default implementation of buffering all the output in a string and passing this to a coderef,
so again beware of the memory consumption if the output is large.

=head1 SEE ALSO

L<Any::Template>, L<Any::Template::Backend>, L<Text::MicroMason>

=head1 VERSION

$Revision: 1.6 $ on $Date: 2005/05/08 18:25:18 $ by $Author: johna $

=head1 AUTHOR

John Alden <cpan _at_ bbc _dot_ co _dot_ uk>

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.

See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt 

=cut

lib/Any/Template.pm
###############################################################################
# Purpose : Consistent interface to templating modules
# Author  : John Alden
# Created : Dec 2004
# CVS     : $Header: /home/cvs/software/cvsroot/any_template/lib/Any/Template.pm,v 1.13 2005/07/21 12:18:40 piersk Exp $
###############################################################################

package Any::Template;

use strict;
use Carp;
use File::Spec;
use File::Find;

use vars qw($VERSION);
$VERSION = sprintf"%d.%03d", q$Revision: 1.13 $ =~ /: (\d+)\.(\d+)/;

sub new {
	my($class, $options) = @_;
	DUMP("Any::Template constructor - options", $options);
	$options->{Options} = {} unless defined ($options->{Options}); #Ensure this key is present

	#Input checking	
	my $backend = $options->{Backend} || $ENV{ANY_TEMPLATE_DEFAULT} 
		or croak("You must nominate a Backend or set the ANY_TEMPLATE_DEFAULT environment variable");
	croak("Package name '$backend' looks incorrect") if($backend =~ /[^\w:]/);
	my @sources = qw(Filename Filehandle String);
	croak("You must supply one of: ".join(", ", @sources)) unless grep {defined $options->{$_}} @sources;
	
	#Load backend
	$backend = join("::", __PACKAGE__, 'Backend', $backend);	
	eval "require $backend";
	die($@) if($@);

	#Create object (containing backend)
	my $self = {};
	$self->{backend} = $backend->new($options);
	DUMP($self);
	return bless($self, $class);
}

sub process {
	my($self, $data, $collector) = @_;
	
	#Input checking
	croak("You must supply a data structure") unless defined($data);
	croak("Data structure should be a reference") unless(ref $data);
	
	my $string;
	if(defined $collector) {
		$self->_process($data, $collector);		
	} else {
		$self->_process($data, \$string);
	}		
	return $string;
}

sub native_object {
	my $self = shift;
	return $self->{backend}->native_object();
}

#Find all the available backends for Any::Template
sub available_backends {
	my @possible_locations = grep {-d $_} map {File::Spec->catfile($_, split(/::/, __PACKAGE__), "Backend")} @INC;
	
	my @backends;
	my $collector = sub {
		return unless $_ =~ /\.pm$/;
		my $file = $File::Find::name;	
		$file =~ s/\Q$File::Find::topdir\E//;
		$file =~ s/\.pm$//;
		my @dirs = File::Spec->splitdir($file);
		shift @dirs;
		$file = join("::", @dirs);
		push @backends, $file;
	};
	
	File::Find::find($collector, @possible_locations);	
	return \@backends;
}

#
# Private functions
#

sub _process {
	my($self, $data, $collector) = @_;
	
	#Preprocess data if required
	$data = $self->{backend}->preprocess($data); 

	#Type-based dispatch
	die("Hash or array refs not supported as sinks") if(ref $collector eq "HASH" || ref $collector eq "ARRAY");
	return $self->{backend}->process_to_string($data, $collector) if(ref $collector eq "SCALAR"); 
	return $self->{backend}->process_to_sub($data, $collector) if(ref $collector eq "CODE");   
	return $self->{backend}->process_to_filehandle($data, $collector) if(ref $collector eq "GLOB"); #Filehandle ref
	return $self->{backend}->process_to_filehandle($data, \$collector) if(ref \$collector eq "GLOB"); #Filehandle
	return $self->{backend}->process_to_file($data, $collector) if(not ref $collector); #Must come after check for glob
	return $self->{backend}->process_to_filehandle($data, $collector); #object - treat as a filehandle
}

#Log::Trace stubs
sub TRACE{}
sub DUMP{}

=head1 NAME

Any::Template - provide a consistent interface to a wide array of templating languages

=head1 SYNOPSIS

	use Any::Template;
	my $template = new Any::Template({
		Backend => 'HTML::Template',
		Filename => 'page.tmpl',
		Options => {'strict' => 0}
	});
	my $output = $template->process($data);

	my $template2 = new Any::Template({
		Backend => 'Text::Template',
		String => $template2_content
	});
	$template->process($data, \*STDOUT);

=head1 DESCRIPTION

This module provides a simple, consistent interface to common templating engines so you can write code that
is agnostic to the template language used in the presentation layer.  This means you can allow your interface 
developers to work in the templating language they're happiest in or write code that works with legacy/in-house 
templating modules but can also be released onto CPAN and work with more standard CPAN templating systems.

By its very nature, this interface only exposes pretty much the lowest common denominator of the template engine APIs.
It does however provide a fairly rich set of input and output mechanisms, using native implementations where available
and providing some default implementations to extend the default set offered by some templating modules.

If you need the quirky features of a particular templating engine, then this may not be for you.
Having said that, in some cases you may be able to encapsulate some of your logic in options passed into the adaptor classes
(either rolling your own adaptors, or improving ours) to pull the relevant strings on the backend module.

Templateing languages supported by backends supplied with this distribution can be found in the README (remember there may be others out there and you can always roll your own).

=head1 METHODS

=over 4

=item my $template = new Any::Template(\%options);

	See below for a list of options

=item $template->process($data_structure, $sink);

	$sink can be:
		- a scalar ref
		- a filename (string)
		- a filehandle (as a glob or glob ref) or an object offering a print method
		- a coderef (output will be passed in as the first argument)

=item $string = $template->process($data_structure);

A convenience form, if no second argument is passed to C<process()>, equivalent to:
	
	my $string;
	$template->process($data_structure, \$string);

except data is passed by value rather than by reference.

=item $templating_engine = $template->native_object();

Allows the native templating engine to be accessed.
This completely breaks the abstraction of Any::Template so it's not recommended you use it
other than as a bridging strategy as part of a refectoring/migration process 
(with a view to ultimately eliminating its use).

=back

=head1 SUBROUTINES

=over

=item available_backends

 $list_of_backends = Any::Template::available_backends();

Scans @INC for a list of modules in the Any::Template::Backend:: namespace.

=back

=head1 ERROR HANDLING

If an error occurs, an exception is raised with die().  You can use an eval block to handle the exception.  $@ will contain an error message.

=head1 CONSTRUCTOR OPTIONS

=over 4

=item Backend

Backends distributed with this module are listed in the distribution README.

See L<Any::Template::Backend> for information on writing your own backends.  
You should be able to create a new backend in a couple of dozen lines of code and 
slot it into the unit test with a one or 2 line change.

=item Filename

	Filename of the template file

=item String

	String containing the template

=item Filehandle

	Reference to a filehandle from which to read the template

=item Options

	A hashref of options passed to the backend templating engine

=back

=head1 ENVIRONMENT

If you don't supply a Backend to the constructor, Any::Template looks for a default Backend in the
ANY_TEMPLATE_DEFAULT environment variable. This allows you to retrofit Any::Template into legacy code without
hard-coding a default templating language or forcing a backwardly-incompatible change to the interface of the code you are retrofitting into.

=head1 CACHING

This module doesn't have built-in caching, however the objects it creates are intended to be cachable
(where possible the backends hold onto precompiled templates that can be fed any number of data structures).
It's therefore up to you what caching strategy you use.  In the spirit of "if you liked this, you might also like..."
L<Cache> and L<Cache::Cache> offer a consistent interface to a number of different caching strategies.

=head1 SEE ALSO

=over 4

=item *

L<Text::MicroMason>

=item *

L<Template>

=item *

L<HTML::Template>
 
=item *

L<Text::Template>

=back 

=head1 VERSION

$Revision: 1.13 $ on $Date: 2005/07/21 12:18:40 $ by $Author: piersk $

=head1 AUTHOR

John Alden <cpan _at_ bbc _dot_ co _dot_ uk>

=head1 COPYRIGHT

(c) BBC 2005. This program is free software; you can redistribute it and/or modify it under the GNU GPL.

See the file COPYING in this distribution, or http://www.gnu.org/licenses/gpl.txt 
 
=cut
